### PR TITLE
More readable project urls

### PR DESCRIPTION
**Pull request by j9peters**
Include slugified project name in project URL's

issue #12 More readable URLs

Co-authored-by: EL246 <25569192+EL246@users.noreply.github.com>

### CODE REVIEW

File: src/main/java/com/recurse/portfolio/web/ProjectController.java [hunk 1]
```diff
@@ -79,29 +81,69 @@ public ModelAndView postNewProject(
...
         @CurrentUser User currentUser,
-        @PathVariable Integer projectId
-    ) {
+        @PathVariable Integer projectId,
+        @PathVariable String projectName
+    ) {     
+        // ensure provided project ID is valid
         Project project = repository.findById(projectId)
-            .orElseThrow(() -> new NotFoundException("project", projectId));
+            .orElseThrow(() -> new NotFoundException("project", projectId));
```
Comment (jasonaowen) [inline src/main/java/com/recurse/portfolio/web/ProjectController.java]: Please watch out for trailing whitespace! You can double check for it with `git grep -E ' +$'`, and/or configure your editor to automatically delete it.

File: src/main/java/com/recurse/portfolio/web/ProjectController.java [hunk 2]
```diff
@@ -175,8 +220,13 @@ public ModelAndView postEditProject(
...
             }
-
-            return new ModelAndView(new RedirectView("/project/" + projectId));
+            
+            // retrieve name of project from model
+            String projectNameFromModel = project.getName();
+            
+            String slugifiedProjectName = slugifyString(projectNameFromModel);          
+            
+            return new ModelAndView(new RedirectView("/project/" + projectId + "/" + slugifiedProjectName));
```
Comment (jasonaowen) [inline src/main/java/com/recurse/portfolio/web/ProjectController.java]: We do this project-with-slug URL construction in a few different places; would you mind extracting a helper method?

File: src/main/java/com/recurse/portfolio/web/ProjectController.java [hunk 3]
```diff
@@ -50,7 +52,7 @@ public ModelAndView getNewProject(
             .addObject("allTags", tagRepository.findAll());
     }
 
-    @PostMapping("/project/")
+    @PostMapping("/project/new")
```
Comment (jasonaowen) [inline src/main/java/com/recurse/portfolio/web/ProjectController.java:57]: I almost did this initially, and probably I should have. Tell me more about your motivation for changing it? Was it for the sake of clarity?

File: src/main/java/com/recurse/portfolio/web/ProjectController.java [hunk 4]
```diff
@@ -79,29 +81,69 @@ public ModelAndView postNewProject(
...
-
-    @GetMapping("/project/{projectId}")
+    
+    // slugify, i.e. create URL-friendly version of, the project name
+    private String slugifyString(String text)
+    {
+        Slugify slugifier = new Slugify();
+        String slugifiedProjectName = slugifier.slugify(text);
+        return slugifiedProjectName;
+    }
```
Comment (jasonaowen) [inline src/main/java/com/recurse/portfolio/web/ProjectController.java]: Making a helper method was a good call, but I think we can make this both more performant and shorter:

If we move the `slugifier` to be an instance variable (`private Slugify slugifier = new Slugify();`), then all that's left is `return slugifier.slugify(text);`, which we could inline.

File: build.gradle [hunk 5]
```diff
@@ -32,6 +32,7 @@ dependencies {
   compile('org.springframework.boot:spring-boot-starter-security')
   compile('org.springframework.boot:spring-boot-starter-thymeleaf')
   compile('org.springframework.boot:spring-boot-starter-web')
+  compile('com.github.slugify:slugify:2.3')
```
Comment (jasonaowen) [inline build.gradle]: This is such a tiny thing, but would you mind keeping this list sorted?

**Review by jasonaowen — CHANGES_REQUESTED** [review]
Thank you so much for working on this, @j9peters and @EL246 ! This is great. :tada: 

I was able to test it locally by creating a new project, editing a project, navigating to various incorrect URLs and getting redirected to the correct one, and verifying that I couldn't see the slug of a project I didn't have permission to view.

There are some code suggestions below. Please also [squash](https://help.github.com/en/articles/about-git-rebase) the merge commit away; let me know if you would like to work together on doing so!

File: src/main/java/com/recurse/portfolio/web/ProjectController.java (hunk 3 above)
Comment (EL246) [reply src/main/java/com/recurse/portfolio/web/ProjectController.java:57]: We noticed the GetMapping and PostMapping URIs for new projects didn't match, so we added _/new_ to the PostMapping URI to keep things consistent.

Comment (jasonaowen) [reply src/main/java/com/recurse/portfolio/web/ProjectController.java:57]: Makes sense, thank you!

**Review by jasonaowen — APPROVED** [review]
Great work, @j9peters and @EL246! Thank you so much.

**Pull request merged**